src/main/java/hudson/plugins/ec2/ssh/EC2MacLauncher.java

```diff
@@ -174,184 +189,287 @@ protected void launchScript(EC2Computer computer, TaskListener listener)
-                    return;
-                }
-                sess.close();
-
-                logInfo(computer, listener, "Creating ~/.hudson-run-init");
+            int bootDelay = node.getBootDelay();
+            if (bootDelay > 0) {
+                logInfo(
+                        computer,
+                        listener,
+                        "SSH service responded. Waiting " + bootDelay + "ms for service to stabilize");
+                Thread.sleep(bootDelay);
+                logInfo(computer, listener, "SSH service should have stabilized");
+            }
 
-                // Needs a tty to run sudo.
-                sess = conn.openSession();
-                sess.requestDumbPTY(); // so that the remote side bundles stdout
-                // and stderr
-                sess.execCommand(buildUpCommand(computer, "touch ~/.hudson-run-init"));
+            // connect fresh as ROOT
+            logInfo(computer, listener, "connect fresh as root");
+            try (ClientSession clientSession = connectToSsh(client, computer, listener, template)) {
+                KeyPair key = computer.getCloud().getKeyPair();
 
-                sess.getStdin().close(); // nothing to write here
-                sess.getStderr().close(); // we are not supposed to get anything
-                // from stderr
-                IOUtils.copy(sess.getStdout(), logger);
+                final boolean isAuthenticated;
+                if (key == null) {
+                    isAuthenticated = false;
+                } else {
+                    clientSession.addPublicKeyIdentity(KeyHelper.decodeKeyPair(key.getKeyMaterial(), ""));
+                    clientSession.auth().await(timeout);
+                    isAuthenticated = clientSession.isAuthenticated();
+                }
+                if (!isAuthenticated) {
+                    logWarning(computer, listener, "Authentication failed");
+                    return; // failed to connect as root.
+                }
 
-                exitStatus = waitCompletion(sess);
-                if (exitStatus != 0) {
-                    logWarning(computer, listener, "init script failed: exit code=" + exitStatus);
-                    return;
+                try (CloseableScpClient scp = createScpClient(clientSession)) {
+                    String timestamp =
+                            Duration.ofMillis(System.currentTimeMillis()).toSeconds() + " 0";
+                    ScpTimestampCommandDetails scpTimestamp =
+                            ScpTimestampCommandDetails.parse("T" + timestamp + " " + timestamp);
+                    String initScript = node.initScript;
+
+                    logInfo(computer, listener, "Creating tmp directory (" + tmpDir + ") if it does not exist");
+                    executeRemote(clientSession, "mkdir -p " + tmpDir, logger);
+
+                    if (initScript != null
+                            && !initScript.trim().isEmpty()
```

Comment:
```suggestion
                    if (StringUtils.isEmpty(initScript) != null
```
Since the project already uses Apache Commons, we can utilize the utility method.

Reply:
This is not related to the `trilead` migration, but I have included it as it is a minor change: fe9a90165a226d3afc7b2ca32128b301ac0719c9